lib/src/rules/avoid_type_to_string.dart

```diff
@@ -0,0 +1,153 @@
+but release-mode compilers often choose to minify these symbols.
+
+**BAD:**
+```
+void bar(Object other) {
+  if (other.runtimeType.toString() == 'Bar') {
+    doThing();
+  }
+}
+
+Object baz(Thing myThing) {
+  return getThingFromDatabase(key: myThing.runtimeType.toString());
+}
+```
+
+**GOOD:**
+```
+void bar(Object other) {
+  if (other is Bar) {
+    doThing();
+  }
+}
+
+class Thing {
+  String get thingTypeKey => ...
+}
+
+Object baz(Thing myThing) {
+  return getThingFromDatabase(key: myThing.thingTypeKey);
+}
+```
+
+''';
+
+// TODO: handle local toString lambda declared in scope
+class AvoidTypeToString extends LintRule implements NodeLintRule {
+  AvoidTypeToString() 
+      : super(
+            name: 'avoid_type_to_string',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry, [LinterContext context]) {
+    final visitor = _Visitor(this);
+    registry.addClassDeclaration(this, visitor);
+    registry.addMixinDeclaration(this, visitor);
+    registry.addExtensionDeclaration(this, visitor);
+    registry.addMethodInvocation(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  InterfaceType thisType;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitClassDeclaration(ClassDeclaration node) {
+    // Nodes visited in DFS, so this will be set before each visitMethodInvocation.
+    thisType = node.declaredElement.thisType;
+  }
+
+  @override
+  void visitMixinDeclaration(MixinDeclaration node) {
+    // Nodes visited in DFS, so this will be set before each visitMethodInvocation.
+    thisType = node.declaredElement.thisType;
+  }
+
+  @override
+  void visitExtensionDeclaration(ExtensionDeclaration node) {
+    // Nodes visited in DFS, so this will be set before each visitMethodInvocation.
+    thisType = node.declaredElement.extendedType as InterfaceType;
+  }
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    final methodName = node.methodName.name;
+    final targetType = (node.realTarget?.staticType is InterfaceType)
+        ?  node.realTarget.staticType as InterfaceType : thisType;
+    final library = node.methodName.staticElement?.library;
+    if (_isToStringOnCoreTypeClass(methodName, targetType, library)) {
+      rule.reportLint(node);
+    }
+
+    node.argumentList.arguments.forEach(_validateArgument);
+  }
+
+  void _validateArgument(Expression expression) {
+    String methodName;
+    InterfaceType targetType;
+    LibraryElement library;
+    if (expression is PropertyAccess) {
+      methodName = expression.propertyName.name;
+      targetType = (expression.realTarget?.staticType is InterfaceType) 
+          ? expression.realTarget?.staticType as InterfaceType : thisType;
+      library = expression.propertyName.staticElement?.library;
+    }
+    else if (expression is SimpleIdentifier) {
```

Comment:
Please run the dart formatter over the code.

Reply:
Oopsies. :)